### ISSUE TITLE
Determining the actual errors on large schema is problematic

### ISSUE THREAD

**Issue by MiYanni** [JamesNK/Newtonsoft.Json.Schema#76]
I am working with a schema that references hundreds of other schemas. To give an idea of scope, when I dereference the main schema using Json.NET Schema, the dereferenced schema file is almost 25 thousand lines.

The situation is such. The JSON I'm validating has 2 actual errors when compared to the schema. If I load this JSON file into Visual Studio, it displays 5 errors. This is because 1 of the 2 errors is nested many schemas deep, so Visual Studio shows errors on each of the objects that are parents to the actual property that has an error. Basically, 1 error shows 4 errors instead of 1.

But, doing this same process with Json.NET Schema, I come up with 753 errors. This is the total number of errors including every child error of every error. After removing duplicates (duplicates are determined by if they have the same line number, line position, and message text), I come up with 98 errors. If I only keep the leaves from the error tree (since `ValidationError` has `ChildErrors`, it is a massive tree), I get 87 errors. 

Now, the top-most count of errors is actually 2. Meaning, `IsValid` returns an `IList` of 2 `ValidationError`. So, that count is correct. The problem is, I want to report the actual error. Meaning, the decedent that is the problem that someone needs to fix. For example, when I fix that problem (the one that generates hundreds of children), I'm left with 1 error that has no children. The error that is left is not nested like this other error, so that's why it doesn't have this issue.

So, how do I go about determining the actual error for the nested error? In investigating the other errors in the leaves (the 87 errors), they are being reported by schemas that are not related at all to the error. It seems like, because of this 1 error, the validator cannot resolve the actual schema that is supposed to be using for the object. 

Is there a possible way to force it to know which schema to validate against? These schemas are build so the Type property of the object is a path to a specific schema. For example, a Type of `Foo/Bar` maps to a URL that contains `Foo.json#/things/Bar`. If I can force it to recognize Type is the intended Schema (instead of trying to let it resolve the OneOf/AnyOf/AllOf), it wouldn't cause this huge mess of errors (I'm hoping).

Any help would be appreciated. Thank you! :-)

**Comment by JamesNK** [comment]
Yes errors do become unwieldy once you have a lot of conditional schemas involved.

> It seems like, because of this 1 error, the validator cannot resolve the actual schema that is supposed to be using for the object.

At any given level of some JSON, with a lot of OneOf/AnyOf/AllOf schemas involved the validator is testing all of them to find the combination that is valid. If it can't find a valid path through them then it has to report all the errors. For example, it doesn't know that in the `oneOf` that failed you wanted your JSON to match the employee schema and not the customer schema, so it returns errors of why both schemas failed.

> Is there a possible way to force it to know which schema to validate against?

That is tricky given how nested JSON and schemas can be. One `ValidationError` there is a `SchemaBaseUri`. That tells you were the schema that generated the error was loaded from. You could use that to filter errors to a specific schema file.

**Comment by MiYanni** [comment]
Awesome. That is a great place to start. The main issue is... for `SchemaBaseUri` to be useful, I need to know information about what is being actually checked in the JSON file. I'm assuming I can use `Path` for this. From what I've read [here](http://www.newtonsoft.com/json/help/html/QueryingLINQtoJSON.htm), I could take the path, parse it, and then use that querying syntax to get the object and properties I need to determine the schema. Does that sound like the right way to do it? The `Path` seems to be the only way to get at what the validator was inspecting in the JSON file at the time the `ValidationError` occurred.

**Comment by JamesNK** [comment]
Yes `Path` is the location in the JSON that the error occurred. You don't need to parse the path, there is `SelectToken` which does it for you already - http://www.newtonsoft.com/json/help/html/SelectToken.htm

**Comment by MiYanni** [comment]
Awesome. I tested it out and it works as well as I can do for now. I noticed that Visual Studio must have quite a bit of smarts in their JSON schema validation, as it seems to give you a 'best guess' if you have a value that doesn't make sense. It must do quite a lot of inference on the JSON and the schema to have those kinds of smarts. I clearly cannot compete with that.

Thanks for your help! I'd suggest adding some 'warning' or extra info about schema error resolution for OneOf/AnyOf/AllOf so that other people know in the future. I'm sure others will have similar questions like mine.
